lib/dry/system/lifecycle.rb

```diff
@@ -57,6 +57,25 @@ def call(*triggers)
       end
 
       # @api private
+      def provides?(component)
+        if component.external?
+          container.source != container && container.source.key?(component.identifier)
```

Comment:
This `container.source != container` confuses me a little. Looking through `LifecycleContainer`, would't `container.source` _always_ `!=` to any other container, because we always create a branch new `Dry::Container.new` whenever a new `LifecycleContainer` is initialised?

Reply:
Also, there's a few places like this where we have to dig through `container` to reach `.source` and ask it questions, like we do here with `container.source.key?`. I'm wondering if we could do a few things to make this a bit better?

- firstly, calling these vars `container` confused me a number of times, because this is different to the normal "container" that dry-system exposes to the outside (which has no `.source` attribute). Perhaps we could pick a better local var name to help distinguish which is which?
- secondly, I wonder if `LifecycleContainer` could offer some extra methods on its own to make it so we don't have to be aware of its internals (i.e. `.source`) whenever we interact with it?

Reply:
@timriley yeah it's messy, I want to refactor this once I test things out in a real app